internal/data/metric.go

```diff
@@ -0,0 +1,637 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package data
+
+import (
+	otlpmetric "github.com/open-telemetry/opentelemetry-proto/gen/go/metrics/v1"
+)
+
+// This file defines in-memory data structures to represent metrics.
+// For the proto representation see https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/metrics/v1/metrics.proto
+
+// MetricData is the top-level struct that is propagated through the metrics pipeline.
+// This is the newer version of consumerdata.MetricsData, but uses more efficient
+// in-memory representation.
+type MetricData struct {
+	resourceMetrics []*ResourceMetrics
+}
+
+func NewMetricData(resourceMetrics []*ResourceMetrics) MetricData {
+	return MetricData{resourceMetrics}
+}
+
+// MetricCount calculates the total number of metrics.
+func (md MetricData) MetricCount() int {
+	metricCount := 0
+	for _, rml := range md.resourceMetrics {
+		metricCount += len(rml.metrics)
+	}
+	return metricCount
+}
+
+func (md MetricData) ResourceMetrics() []*ResourceMetrics {
+	return md.resourceMetrics
+}
+
+func (md MetricData) SetResource(r []*ResourceMetrics) {
+	md.resourceMetrics = r
+}
+
+// A collection of metrics from a Resource.
+//
+// Must use NewResourceMetrics functions to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type ResourceMetrics struct {
+	// The resource for the metrics in this message.
+	// If this field is not set then no resource info is known.
+	resource *Resource
+
+	// A list of Spans that originate from a resource.
+	metrics []*Metric
+}
+
+func NewResourceMetrics(resource *Resource, metrics []*Metric) *ResourceMetrics {
+	return &ResourceMetrics{resource, metrics}
+}
+
+func (rm *ResourceMetrics) Resource() *Resource {
+	return rm.resource
+}
+
+func (rm *ResourceMetrics) SetResource(r *Resource) {
+	rm.resource = r
+}
+
+func (rm *ResourceMetrics) Metrics() []*Metric {
+	return rm.metrics
+}
+
+func (rm *ResourceMetrics) SetMetrics(s []*Metric) {
+	rm.metrics = s
+}
+
+// Metric defines a metric which has a descriptor and one or more timeseries points.
+//
+// Must use NewMetric* functions to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type Metric struct {
+	// Wrap OTLP Metric.
+	orig *otlpmetric.Metric
+
+	// Override a few fields. These fields are the source of truth. Their counterparts
+	// stored in corresponding fields of "orig" are ignored.
+	metricDescriptor    MetricDescriptor
+	int64DataPoints     []*Int64DataPoint
+	doubleDataPoints    []*DoubleDataPoint
+	histogramDataPoints []*HistogramDataPoint
+	summaryDataPoints   []*SummaryDataPoint
+}
+
+func NewMetric() *Metric {
+	return &Metric{orig: &otlpmetric.Metric{}}
+}
+
+// NewMetricSlice creates a slice of Metrics that are correctly initialized.
+func NewMetricSlice(len int) []Metric {
```

Comment:
not clear why `[]Metric` and not `[]*Metric`, in contrast `NewMetric()` creates a reference, perhaps worth a comment.

Reply:
Will fix later